### PR TITLE
Template parameter {usage} for XYZ raster and vector tiles

### PR DESCRIPTION
**Pull request by vcloarec**
For XYZ tile-based raster layers and for vector tile layers, this PR allows the possibility to optionally have {usage} template parameter (in addition to {x}, {y}, {z}) specified. If present, it will be replaced by the context in which the request is being made. The idea is that MapTiler needs to recognize the context in which tiles are used and bill users accordingly. 

There are these contexts recognized:
- “view” - when the request comes from 2D/3D map canvas
- “export” - when the request comes from export or print layout for
high-quality output (map preview in print layout GUI will still count
as “view”)
- “unknown” - when coming from some other context, but the source is
unknown

Funded by [MapTiler](https://www.maptiler.com/)

### CODE REVIEW

File: src/3d/terrain/qgsterraintexturegenerator_p.cpp [hunk 1]
```diff
@@ -159,6 +159,7 @@ QgsMapSettings QgsTerrainTextureGenerator::baseMapSettings()
   mapSettings.setFlag( Qgis::MapSettingsFlag::Render3DMap );
   mapSettings.setTransformContext( mMap.transformContext() );
   mapSettings.setPathResolver( mMap.pathResolver() );
+  mapSettings.setRendererUsage( Qgis::RendererUsage::View );
```
Comment (nyalldawson) [inline src/3d/terrain/qgsterraintexturegenerator_p.cpp:162]: What about when a 3d map is being exported via layouts?

File: src/core/qgis.h [hunk 2]
```diff
@@ -1250,6 +1250,22 @@ class CORE_EXPORT Qgis
     };
     Q_ENUM( AngularDirection )
 
+    /**
+     *  Usage of the renderer.
+     *
+     *  \note This usage not alter how the map gets rendered but the intention is that data provider
```
Comment (nyalldawson) [inline src/core/qgis.h:1256]: Can you move this note to the QgsRenderContext/QgsMapSettings members instead? It's not a generic property of the enum itself, rather how it's interpreted in the rendering.

File: src/core/qgsmapsettings.h [hunk 3]
```diff
@@ -805,6 +805,22 @@ class CORE_EXPORT QgsMapSettings : public QgsTemporalRangeObject
...
+     */
+    Qgis::RendererUsage rendererUsage() const;
+
+    /**
+     * Sets the rendering usage
+     *
+     * \see rendererUsage()
+     * \since QGIS 3.24
+     */
+    void setRendererUsage( const Qgis::RendererUsage &rendererUsage );
```
Comment (nyalldawson) [inline src/core/qgsmapsettings.h:822]:
```suggestion
    void setRendererUsage( Qgis::RendererUsage rendererUsage );
```

File: src/core/raster/qgsrasterlayerrenderer.cpp [hunk 4]
```diff
@@ -282,6 +282,8 @@ QgsRasterLayerRenderer::QgsRasterLayerRenderer( QgsRasterLayer *layer, QgsRender
   }
 
   mClippingRegions = QgsMapClippingUtils::collectClippingRegionsForLayer( *renderContext(), layer );
+
+  mFeedback->setRendererUsage( rendererContext.rendererUsage() );
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayerrenderer.cpp:286]: This is ugly. We may as well just make a "setRenderContext"/"renderContext" member in QgsRasterBlockFeedback rather then add pieces of render context bit by bit.

**Comment by nyalldawson** [comment]
I have a general issue with using the "{usage}" token here. I understand the use case for this is heavily maptiler dependant, but "usage" is a very generic term. I can see that potentially in future we'll have an issue if users have designed a setup which uses the "usage" token as a way of recording local usage in non-maptiler contexts, and then maptiler decides that the existing view/export values need updating/changing, potentially then breaking the non-maptiler usage of this. 

I'd suggest we add two different tokens. Keep "usage" as you've done here as a generic, non-maptiler specific token, and then add another token "maptiler_usage" which can be used specifically to meet the maptiler requirements.

The various xyz/vector tiles configuration widgets should also ideally be updated to add some inline help text describing the available tokens.

**Comment by wonder-sk** [comment]
Hi @nyalldawson the `{usage}` token is not meant to be specific to maptiler... we have decided on the naming `{usage}` and `view` / `export` / `unknown` values ourselves, hoping that they are good in other (non maptiler) contexts. We are free to pick whatever convention works best for us, so there's no need to have a generic `{usage}` and then maptiler specific `{maptiler_usage}`. As far as I understand, maptiler team is flexible and would go with whatever we implement.

Do you have any preference how to change the naming of usage contexts, or maybe even increase the number of recognized contexts?

To further explain the rationale: some vendors of data (especially aerial imagery) require that downstream providers keep track of contexts in which their data are used (whether for on-screen display or for printing) - without that maptiler team would not be legally allowed to make the datasets available in QGIS.

File: src/3d/terrain/qgsterraintexturegenerator_p.cpp (hunk 1 above)
Comment (vcloarec) [reply src/3d/terrain/qgsterraintexturegenerator_p.cpp]: very good point, that was not treated...
Fixed with https://github.com/qgis/QGIS/pull/46731/commits/5a5bd20fde24867eca9b7c4d15ee56df0071379a but I don't like that we need to reset the engine to force to render again.

File: src/core/qgsmapsettings.h [hunk 5]
```diff
@@ -805,6 +805,25 @@ class CORE_EXPORT QgsMapSettings : public QgsTemporalRangeObject
...
+    /**
+     * Sets the rendering usage
+     *
+     * \note This usage not alter how the map gets rendered but the intention is that data provider
+     * knows the context of rendering and may report that to the backend.
+     *
+     * \see rendererUsage()
+     * \since QGIS 3.24
+     */
+    void setRendererUsage( const Qgis::RendererUsage &rendererUsage );
```
Comment (nyalldawson) [inline src/core/qgsmapsettings.h:825]:
```suggestion
    void setRendererUsage( Qgis::RendererUsage rendererUsage );
```

**Comment by nyalldawson** [comment]
@wonder-sk 
> As far as I understand, maptiler team is flexible and would go with whatever we implement.

So we'd be fine to change these values future (e.g. extending them with new items, such as "server") without impacting maptiler?

**Comment by wonder-sk** [comment]
> So we'd be fine to change these values future (e.g. extending them with new items, such as "server") without impacting maptiler?

Yeah, that's correct

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**